xonsh/tools.py

```diff
@@ -805,6 +805,21 @@ def swap(namespace, name, value, default=NotImplemented):
     else:
         setattr(namespace, name, old)
 
+
+@contextlib.contextmanager
+def swap_values(d, updates, default=NotImplemented):
+    """Updates a dictionary (or other mapping) with values from another mapping,
+    and then restores the original mapping when the context is exited.
+    """
+    old = {k: d.get(k, NotImplemented) for k in updates}
```

Comment:
picked that up from a python core dev talk (iirc Raymond Hettinger on idiomatic python)
and kind of took it for granted, no idea if there is anything on the official docs.